codes/pop.py

```diff
@@ -0,0 +1,209 @@
+        message = f"Population not found for adm{adm_level}_name: {null_adm}"
+        if errors=="warn":
+            warnings.warn(message)
+        else:
+            raise ValueError(message)
+
+    return policies
+
+def merge_policies_with_population(policies, country_code, max_adm_level, errors='raise'):
+    """Assign all populations at all adm-levels to DataFrame of policies
+
+    Args:
+        policies (pandas.DataFrame): List of policies as formatted in ```data/raw/{`country_code`}/{`country_code`}_policy_data_sources.csv```
+        country_code (str): Three-letter country code of requested populations
+        max_adm_level (int): Adm-level at and below which populations should be assigned, down to adm1
+            e.g. `max_adm_level` == 3 would assign populations at adm-levels 1, 2, and 3
+    Kwargs:
+       errors (str): Error-handling behavior. Options are "raise", "ignore", "warn", and "error" (default)
+
+    Returns:
+        pandas.DataFrame: `policies` with new columns, "adm_{`adm_level`}_pop" for each `adm_level` from 1 to `max_adm_level`
+
+    """
+    for adm_level in range(1, max_adm_level + 1):
+        policies = merge_policies_with_population_on_level(policies, adm_level, country_code, errors)
+
+        # Check that all non-"All" populations are assigned
+        condition = policies.loc[policies[f"adm{adm_level}_name"] != "All", f"adm{adm_level}_pop"].isnull().sum() == 0
+        if errors=="raise":
+            assert condition
+        elif errors=="ignore":
+            pass
+        elif not condition:
+            null_adm = sorted(set(policies.loc[policies[f"adm{adm_level}_pop"].isnull(), f'adm{adm_level}_name']) - set(['All']))
+            message = f"Population not found for adm{adm_level}_name: {null_adm}"
+            if errors=="warn":
+                warnings.warn(message)
+            else:
+                raise ValueError(message)
+
+    return policies
+
+def merge_cases_with_population_on_level(epi_df, adm_level, country_code, errors='raise'):
+    """Assign all populations at a given adm-level to DataFrame of epidemiological (cases) data
+
+    Args:
+        epi_df (pandas.DataFrame): List of cases as formatted in ```data/processed/{`adm_level`}/{`country_code`}_processed.csv```
+        adm_level (int): Adm-level of requested populations.
+        country_code (str): Three-letter country code of requested populations
+    Kwargs:
+       errors (str): Error-handling behavior. Options are "raise", "ignore", "warn", and "error" (default)
+
+    Returns:
+        pandas.DataFrame: `epi_df` with a new column, "population"
+
+    """
+    adm_pops = get_adm_pops(adm_level, country_code)
+    result = pd.merge(
+        epi_df, 
+        adm_pops, 
+        how='left',
+        right_index=True, 
+        left_on=get_adm_fields(adm_level)
+    ).rename(
+        columns={f'adm{adm_level}_pop':'population'}
+    )
+
+    # Check that all non-"All" populations are assigned
```

Comment:
see comments above. We should turn this into a function called `check_pop_merge` or something